022-feature-gates.md

```diff
@@ -0,0 +1,125 @@
+
+#### Introducing experimental or unstable APIs
+
+Experimental features or APIs might benefit from the feature gate maturity model.
+They might start as alpha features and be disabled by default.
+During this stage, they can be evaluated and tested.
+Features in the alpha state might be seen as experimental because of being considered risky (with high probability of bugs and issues) or because the feature / behavior might change in the future.
+As the new feature matures and becomes less experimental, it might move to beta and be enabled by default..
+Finally, when the feature is seen as stable and ready for _production_, the feature gate should move to GA and possibly removed / always enabled.
+
+An example of this use-cases might be for example using a feature gate for the StatefulSet removal or possibly some features related to KIP-500 / ZooKeeper removal.
+
+#### Enabling / disabling a feature
+
+In some cases, different users might have different preferences and it might be hard to find a middle ground which would make everyone happy and satisfied.
+In such case, the feature gates might be used to modified the behavior of the operator and enable / disable certain features.
+In use case like this, the feature gate might not strictly follow the maturity model.
+It might be disabled by default even when in the GA maturity stage - the GA stage would in this case indicate that the feature is stable and should not change.
+In some cases it might also make sense for these features to be introduced directly in the GA stage - skipping the alpha and beta stages.
+Feature gates used for this use cases might be never always enabled and removed.
+
+An example of this use case might be feature gates to disable automatic creation of Pod Disruption Budgets or Network Policies.
+
+### Implementation
+
+Initially the feature gates support should be implemented in Cluster Operator.
+But the same logic can be also applied to User and Topic Operators.
+The feature gates will be configured in an environment variable `STRIMZI_FEATURE_GATES`.
+They should be configures as a comma separated list of feature gate names prefix with `+` for enabling or `-` for disabling the feature gate.
+For example `-PodDisruptionBudgets,-NetworkPolicies,+ControlPlaneListener`.
+If any feature gate appears in the list more then once, the operator will throw an exception and exit.
+The operator will throw an exception and exit also in case a non-existing or removed feature gate appears in the list.
+Feature gates not in the list will have the default value.
+
+The feature gates will be parsed and stored in the `ClusterOperatorConfig` class / object.
+The `ClusterOperatorConfig` object is already used in different places in the operator and makes the feature gates easily accessible without any major changes to the code structure.
+
+## Compatibility
+
+There are no compatibility issues.
```

Comment:
I know this is just a proposal, but I wonder how we plan to have feature gated CRD API changes, i.e. if we add a new experimental/alpha field, say, `spec.alpha` , how do we ensure this doesnt make it into the GA release?
Are we going to maintain multiple CRDs?
Or simply just allow devs to overwrite the CRD locally? (I think the latter is ill advised, but the former sounds like more generated file checkins!)

Reply:
I do not think we can do it as that would be:
* Breaking the compatibility
* Forcing the people to often convert the resources which we want to avoid

The main idea of the feature gates for me is to avoid the CRD changes where not needed to control things. But if you really need to change the API, I'm not sure you need the feature gate. E.g. when you need to add new field to the CRD to configure something, do you really need the feature gate to control it? Or is either having it or not having it in the CR enough? At the end, the maturity level can be carried also in the API reference documentation.

Reply:
Ah that makes a lot of sense, apologies I was still viewing it from the old way, but indeed, if there is new config sections, I suppose we can include the new fields and just ignore doing anything with them if certain features not enabled.